### DISCUSSION TITLE
Textmate thread safe

### DISCUSSION THREAD

**Discussion by spf-oss** [danipen/TextMateSharp#25] [Q&A] [answered]
Calling "grammar.TokenizeLine(line,rulestack)" by two threads at the same time will cause thread safety problems. How do you solve them?

**Comment by danipen** [comment, marked as answer]
Each AvaloniaEdit instance [uses](https://github.com/AvaloniaUI/AvaloniaEdit/blob/a88f8f8d2a21ecd09c933fd1a1ce0f7b1a4c63de/src/AvaloniaEdit.TextMate/TextMate.cs#L33) one intance of the [TMModel](https://github.com/danipen/TextMateSharp/blob/master/src/TextMateSharp/Model/TMModel.cs) class to manage multithreading. 

It launches only one thread per grammar.

**Comment by spf-oss** [comment]
If I want to achieve the effect of vscode, when the code is very long, first of all, one thread of vscode will refresh the code color from the code in the visual area (the color may be wrong at this time), and then another thread will refresh the color from changing the position. At this time, there will be a phenomenon that the syntax starts two threads.

In this way, you can achieve the rapid response of code color and ensure the correctness after refreshing.

What should I do to achieve this effect?

**Comment by danipen** [comment]
Well this behavior is [already implemented in AvaloniaEdit](https://github.com/AvaloniaUI/AvaloniaEdit/blob/a88f8f8d2a21ecd09c933fd1a1ce0f7b1a4c63de/src/AvaloniaEdit.TextMate/TextEditorModel.cs#L80
) and it's working fine. Take a look.

It parses the viewport in the UI thread and starts parsing the document in a bg thread. Every time the viewport is scrolled, the view port is re-parsed.
